src/visGeometry/ColorHandler.ts

```diff
@@ -0,0 +1,206 @@
+        }
+
+        return new Color(
+            this.colorsData[colorId * 4],
+            this.colorsData[colorId * 4 + 1],
+            this.colorsData[colorId * 4 + 2]
+        );
+    }
+
+    private addNewColor(color: number | string): number {
+        const colorNumber = convertColorStringToNumber(color);
+        const newColor = [
+            ((colorNumber & 0x00ff0000) >> 16) / 255.0,
+            ((colorNumber & 0x0000ff00) >> 8) / 255.0,
+            ((colorNumber & 0x000000ff) >> 0) / 255.0,
+            1.0,
+        ];
+        const colorDataIndex = this.getColorDataIndex(newColor);
+        if (colorDataIndex !== -1) {
+            // found the color, need to return the colorId to the
+            // external caller, with no other changes needed
+            return this.convertDataColorIndexToId(colorDataIndex);
+        }
+        // the color isn't in colorsData, so add it and return the colorId
+        const newColorDataIndex = this.colorsData.length;
+        const newArray = [...this.colorsData, ...newColor];
+        const newColorData = new Float32Array(newArray.length);
+        newColorData.set(newArray);
+        this.colorsData = newColorData;
+        return this.convertDataColorIndexToId(newColorDataIndex);
+    }
+
+    /**
+     * Check if the new color is in our current array of color options, if not,
+     * add it before returning the index
+     */
+    private getColorId(color: string): number {
+        return this.addNewColor(color);
+    }
+
+    public updateColorArray(colors: (number | string)[]): {
+        colorArray: Float32Array;
+        numberOfColors: number;
+    } {
+        this.colorsData = this.setColorArray(colors);
+        return {
+            colorArray: this.colorsData,
+            numberOfColors: this.numberOfColors,
+        };
+    }
+
+    public clearColorMapping(): void {
+        this.idColorMapping.clear();
+    }
+
+    public setColorForAgentTypes(
+        agentTypes: number[],
+        color: string | number
+    ): { colorArray: Float32Array; numberOfColors: number } {
+        const colorString = convertColorNumberToString(color);
+        const colorId = this.getColorId(colorString);
+        /**
+         * Sets one color for a set of ids, using an index into a color array
+         * @param ids agent ids that should all have the same color
+         * @param colorId index into the numberOfColors
+         */
+        agentTypes.forEach((id) => {
+            this.setColorForAgentType(id, colorId);
+        });
+        return {
+            colorArray: this.colorsData,
+            numberOfColors: this.numberOfColors,
+        };
+    }
+
+    public getColorForAgentType(agentType: number): Color {
+        const index = this.getColorIdForAgentType(agentType);
+        return this.getColorById(index);
+    }
+
+    public getColorInfoForAgentType(agentType: number): AgentColorInfo {
+        const color = this.getColorForAgentType(agentType);
```

Comment:
Nit: getColorForAgentType also calls getColorIdForAgentType, and then you call it again on the next line.. so maybe there is a small refactor/cleanup to remove this redundancy. Not sure tho.

Reply:
yeah, they were essentially the same function with one returning less data. I changed the places were the parent module only wanted the color (and not the colorId) to just unpack the one variable from `AgentColorInfo`